nats/src/main.rs

```diff
@@ -91,13 +98,57 @@ impl ConnectionConfig {
     }
 }
 
+/// NATSClientBundles hold a NATS client and information (subscriptions)
+/// related to it.
+///
+/// This struct is necssary because subscriptions are *not* automatically removed on client drop,
+/// meaning that we must keep track of all subscriptions to close once the client is done
+struct NATSClientBundle {
+    pub client: async_nats::Client,
+    pub subscriptions: Vec<NATSSubscriberBundle>,
+}
+
+impl NATSClientBundle {
+    fn new(
+        client: async_nats::Client,
+        subscriptions: Option<Vec<NATSSubscriberBundle>>,
+    ) -> NATSClientBundle {
+        NATSClientBundle {
+            client,
+            subscriptions: subscriptions.unwrap_or(Vec::new()),
+        }
+    }
+}
+
+/// NATSSubscriberBundles hold a NATS subscriber and
+/// information related to the subscriber
+///
+/// As connecting with a subscriber does not save the subject on which you connected
+struct NATSSubscriberBundle {
+    pub subject: String,
+    pub ctrl_tx: Option<mpsc::Sender<NATSLoopControlMessage>>,
```

Comment:
If we do go down this route, a [`broadcast` channel](https://docs.rs/tokio/latest/tokio/sync/index.html#broadcast-channel) with a capacity of one is probably the best bet (`oneshot` only allows a single consumer). Then you can clone the receiver multiple times and send the shutdown once to everyone)

Reply:
Right now I'm using a basic `mpsc::channel` -- since there's only one listen loop that's what I need right? one receiver, and multiple senders that could make the rx loop stop.

Reply:
removed in favor of holding `JoinHandle`s